src/shell.ts

```diff
@@ -0,0 +1,170 @@
+import { T7Map } from './map'
+import { Terminal } from '@tuzig/xterm'
+import { loadCommands, Command } from './commands'
+import { Fields, Form } from './form'
+import { Clipboard } from "@capacitor/clipboard"
+
+export class Shell {
+
+    prompt = "T7> "
+
+    map: T7Map
+    t: Terminal
+    onKey: (ev: KeyboardEvent) => void
+    active = false
+    activeForm: Form | null
+    commands: Map<string, Command>
+    field = ''
+
+    constructor(map: T7Map) {
+        this.map = map
+        this.t = map.t0
+    }
+
+    start() {
+        if (this.active)
+            return
+        this.active = true
+        this.commands = loadCommands(this)
+        let field = ''
+        this.t.scrollToBottom()
+        this.onKey = async ev => {
+            const key = ev.key
+            switch (key) {
+                case "Enter":
+                    this.t.write("\n")
+                    await this.handleLine(field)
+                    field = ''
+                    break
+                case "Backspace":
+                    if (field.length > 0) {
+                        field = field.slice(0, -1)
+                        this.t.write("\b \b")
+                    }
+                    break
+                default:
+                    if (key.length == 1) { // make sure the key is a char
+                        field += key
+                        this.t.write(key)
+                    }
+            }
+            this.field = field
+        }
+    }
+
+    async handleLine(input: string) {
+        const [cmd, ...args] = input.trim().split(/\s+/)
+        await this.execute(cmd, args)
+        this.field = ''
+        this.printPrompt()
+    }
+
+    async execute(cmd: string, args: string[]) {
+        if (this.activeForm) 
+            this.escapeActiveForm()
+        if (!cmd)
+            return
+        const command = this.commands.get(cmd)
+        if (!command)
+            return this.t.writeln(`Command not found: ${cmd}`)
+        this.active = false
+        await command.execute(args)
+        this.active = true
+    }
+
+    async runCommand(cmd: string, args: string[]) {
+        this.escapeActiveForm()
+        this.map.interruptTTY()
+        this.field = [cmd, ...args].join(' ')
+        this.printPrompt()
+        this.t.write("\n")
+        await this.execute(cmd, args)
+        this.field = ''
+        this.printPrompt()
+    }
+
+    async newForm(fields: Fields, type: "menu" | "choice" | "text", title="") {
```

Comment:
newABD is used for functions that allocate and initialize a certain object. In this case we're doing that but also running the form. You can either change the name to `runForm` or leave the `newForm` move the calls to `menu`|`chooseFields`|`start` to the caller function